File: Java/PG/src/순위.java
import java.io.IOException;
import java.util.Arrays;

public class 순위 {
    public static int solution(int n, int [][] results){
        int answer = 0;
        int min = 5000;
        int [][] dp = new int [n+1][n+1];
        for(int x = 0; x < n+1; x++){
            Arrays.fill(dp[x], min);
        }
        for(int [] result : results){
            dp[result[0]][result[1]] = 1;
        }
        for(int z = 1; z < n+1; z++){
            for(int x = 1; x < n+1; x++){
                for(int y = 1; y < n+1; y++){
                    if(dp[x][y] > dp[x][z] + dp[z][y]){
                        dp[x][y] = dp[x][z] + dp[z][y];
                    }
                }
            }
        }
        for(int x = 1; x < n+1; x++){
            boolean flag = true;
            for(int y = 1; y < n+1; y++){
                if(x == y){
                    continue;
                }
                if( dp[x][y] == min && dp[y][x] == min){
                    flag = false;
                    break;
                }
            }
            if(flag){
                answer += 1;
            }
        }
        return answer;

    }
    public static void main(String[] args) throws IOException{
        System.out.println(solution(5, new int [][] {{4,3},{4,2},{3,2},{1,2},{2,5}}));
    }
}

File: Test/순위.js
solution(5,[[4, 3], [4, 2], [3, 2], [1, 2], [2, 5]]);
function solution(n, results){
    let answer = 0;
    let dp = Array.from(Array(n+1), () => Array(n+1).fill(Infinity));
    for(let [winner, loser] of results){
        dp[winner][loser] = 1;
    }
    for(let z = 1; z < n+1; z++){
        for(let x = 1; x < n+1; x++){
            for(let y = 1; y < n+1; y++){
                if(dp[x][y] > dp[x][z] + dp[z][y]){
                    dp[x][y] = dp[x][z] + dp[z][y];
                }
            }
        }
    }
    for(let x = 1; x < n+1; x++){
        let flag = false;
        for(let y = 1; y < n+1; y++){
            if(x === y){
                continue;
            }else{
                if(dp[x][y] == Infinity && dp[y][x] == Infinity){
                    flag = true;
                    break;
                }
            }
        }
        if(!flag){
            answer += 1;
        }
    }
    console.log(dp);
    console.log(answer);
    const a = Number.MAX_SAFE_INTEGER;
    console.log(a);
    const b = a+1;
    console.log(b);
    console.log(a+a);
}